pkcs11: leverage sc_wait_for_event to avoid unneccesary queries to card

frankmorgner opened the pull request:
card_detect_all() is now used as global entry point even to reader specific card detection (card_detect() is removed). This enforces a constant update of reader_states so that no events are silently dropped. This removes the need for having a timer in C_GetSlotInfo() for avoiding too many queries. If the reader implementation (i.e. reader-*.c) does not implement wait_for_event, then all slots will be queried on request just like it used to be.

fixes https://github.com/OpenSC/OpenSC/issues/3107

<!--
Thank you for your pull request.

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX'
(without quotes) in the commit message.

Mention which card(s) are used during testing. To get the name of your card,
run this command: `opensc-tool -n`
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PKCS#11 module is tested
- [ ] Windows is tested (pkcs11-tool with --test-hotplug, --test, --test-threads)
- [ ] Linux is tested (pkcs11-tool with --test-hotplug, --test, --test-threads)
- [ ] macOS is tested (pkcs11-tool with --test-hotplug, --test, --test-threads)

dengert commented:
Cool.

Jakuje commented:
Code wise I do not see any issue, but some tests look like failing (even after the rerun of the previous failure) so it clearly needs some more investigation and testing.

xhanulik commented:
@frankmorgner What is the state of this PR? Do we want this to go into 0.26.0?

frankmorgner commented:
This is a critical step, I would like to see this tested on the different platforms (i.e. PC/SC providers). I have extended the above check list accordingly. As I don't currently have much time for testing, I don't believe this will be ready in time for 0.26.0, even though I would love to see this included...

Jakuje commented on src/libopensc/reader-pcsc.c:
```diff
@@ -101,10 +101,20 @@ void APDU_LOG(u8 *rbuf, uint16_t rsize)
 #endif
 
+#define MAX_PCSC_EVENT_LISTENERS
+#define ARRAY_SIZE(x) (sizeof(x)/sizeof(x[0]))
```
we have this helper already in `pkcs11-tool.c`. Would it make sense to move it to some common header file?

Jakuje commented on src/libopensc/reader-pcsc.c:
```diff
@@ -101,10 +101,20 @@ void APDU_LOG(u8 *rbuf, uint16_t rsize)
+	 * slots to allow pcsc_cancel to abort each blocking operation. PC/SC
+	 * events are tied to the SCARDCONTEXT, which is why we cannot reuse
+	 * pcsc_ctx for watching events as that may interfere with its usage of
+	 * in `refresh_attributes()`, that may swollow events from a different
```
```suggestion
	 * in `refresh_attributes()`, that may swallow events from a different
```

Jakuje reviewed (commented):
thank you for the work on this! Only nits commented inline. We also need to figure out of the isoapplet failure is relevant to these changes or to something else.

frankmorgner committed:
pkcs11: add PC/SC status to each slot to detect changes

This detects changes of reader/card without actually querying the device
and avoids IO latencies.

frankmorgner commented:
I think the implementation is finally working as intended. Card detection is only carried out, when there actually has been some reader/card event on the PC/SC level, otherwise no APDU is sent.

What is strange is that *only* the oseid tests don't work as expected and they are failing in the exact way as the CI tests of #3620 for oseid are failing. This is still the case, when I [add debugging via pcscd (2c70cef)](https://github.com/OpenSC/OpenSC/pull/3126/commits/2c70cef2dfbdb311487eb4f2a34751e17d941acb) (the log of the tests can be found [here](https://productionresultssa5.blob.core.windows.net/actions-results/59f7bc8e-d8d2-4cfa-a170-7e1571971cd4/workflow-job-run-51ea3a2f-a1a5-55fa-918b-ce636212d706/logs/job/job-logs.txt?rsct=text%2Fplain&se=2026-04-15T20%3A17%3A03Z&sig=BvEDats8CJ0sP2EgeUp5%2FlYSOLKijr1IqaLRvsToGIY%3D&ske=2026-04-15T21%3A39%3A05Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2026-04-15T17%3A39%3A05Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-11-05&sp=r&spr=https&sr=b&st=2026-04-15T20%3A06%3A58Z&sv=2025-11-05)). However, if I [add debugging of OpenSC (cd687c0)](https://github.com/OpenSC/OpenSC/pull/3126/commits/cd687c056fd08f3201f274edb5697daeff20ae41), i.e. the head of the current pull request, then oseid fails with a [different error](https://github.com/OpenSC/OpenSC/actions/runs/24479701794/job/71549961112?pr=3126). The problem(s) always happen in the in the symmetric encryption/decryption test.

The fact, that both #3620 and this pull request #3126 are failing with the same error only for oseid despite a different implementation approach lets me assume that this may be related to some bug inside the oseid simulator. @popovec , could you check if you can maybe spot some hints in the linked logs?

frankmorgner committed:
oseid: disable broken encryption test case

see https://github.com/OpenSC/OpenSC/pull/3711

Jakuje reviewed (approved):
Thank you for taking time to get this done! Change looks good to me. CI looks happy (except for known failures).
Should we aim to get this into the next release?

frankmorgner commented:
Yes, perfect. It is good to be included
